### DISCUSSION TITLE
Multiple Threads to the Same Device

### DISCUSSION THREAD

**Discussion by bgile** [nornir-automation/nornir#822] [General]
This came up in a side conversation here: (https://github.com/nornir-automation/nornir/issues/325)
I'm not sure what ever came of this.  Personally, I would love to be able to do this.  My use case is AdTran devices which have no API, basically, all I can do is login and use the CLI.  They have a max login of 8 so if I can send say 5 concurrent sessions and interact with the CLI it would greatly cut down on my automation work.

If anyone else has written something to bypass this I would love to see what they have come up with.  Short of writing my own threading and thread workers etc... I was hoping to keep this all inside the frameworks capibility.

Thanks,

**Comment by ktbyers** [comment]
@bgile Just enter multiple `names` in Nornir inventory and point it to the same device. You would need to coordinate the work being done using each name however in the Nornir-Python code that you write.

Whether `AdTran` will actually support this or not...you would have to test. It would not surprise me if there are limitations here on the device side.

**Reply by bgile** [reply]
@ktbyers Are you just talking about multiple device entries for the same device name?  I'm pretty sure I have tried that before and it did not work, but I will try again to see if I was mistaken/forgot.

**Reply by ktbyers** [reply]
In nornir create two different entries that point to same underlying IP address/hostname:

```yaml
---
# hosts.yaml file
device1:
    hostname: my_device.domain.com
    groups:
      - ios

device2:
    hostname: my_device.domain.com
    groups:
      - ios
```

If that doesn't work, you will probably need to clarify what the failure is (i.e. why it is not working).

**Reply by bgile** [reply]
Ok sounds good.  I will give that a shot.  previously I belive the "device1" would have just been the same, but instead I will make some false names that just point to the same domain and test.

**Reply by bgile** [reply]
Ok tested it out and it totally works.  As long as the host key was unique I can log into the same domain multiple times at once.  This is great!  Thanks for the help :)

**Reply by bgile** [reply]
Well, I was excited till I realized I was not articulating what I was really after correctly haha.  Yes, I needed multiple logins to the same device, but also needed to send different commands.  The thought would be to increase iteration time by being able to send different commands per device, i.e. device1 sends x and device2 sends y but they would go out at the same time.  This is not how the framework behaves, I would probably need to use async or something.?

**Reply by ktbyers** [reply]
Doesn't have anything to do with async vs threads. You need to figure out the logic of how to send cmd1 to device1 and cmd2 to device2. Should be doable.

**Reply by dbarrosop** [reply]
Assign them to different groups and run cmd1 to one group and cmd2 to the other group

**Reply by bgile** [reply]
Well, I thought I would update for what it's worth.  I made 4 threads that pull from a queue that contains all the commands I want to execute.  Inside my job function for each worker, I filter each device in the inventory with a max of 4 devices (all pointing to the same domain as discussed above) so that each command execution from the queue is its own session and then I basically just recycle the devices as one completes so I never clobber the other.  Execution seems to work great, I will run 4 sessions to the same device and execute 4 different commands and they return at the same time and the next 4 commands get picked up, and so on.

Did I do this the best way?  Not sure haha, but it seems to work very well so far in testing.

If you guys ever think about putting in async support I think that would be great :)  As always I love this framework, it has allowed me to do a lot.

**Reply by bgile** [reply]
@dbarrosop Sorry did not see your suggestion.  
I see, makes sense to use the group as a way to differentiate.  I will test that out as well.  Thanks all.
